### PR TITLE
Python

### PR DESCRIPTION
**Pull request by derekendres**

### CODE REVIEW

File: basic_samples/Complex_Dynamic/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+This sample sends OMF messages that are Dynamic (meaning the data is indexed) and Complex (meaning more than 1 value per index).  The data is a temperature and pressure reading at a timestamp.  The sample creates the OMFType, OMFContainer and then waits for keyboard input for the data to send.  The input is the temperature and pressure, with time being the time it is received.  After the initial one-time creation of the type and container, the application is event-driven.  Entries can also be sent as run time parameters
+
+### Test Details
+
+The test checks to make sure that the OMF messages were properly handled by the appropriate OMF endpoint.  It leverages that entries can be entered at runtime.  It then cleans up endpoints and confirms 
+
+---
+ 
+For the OMF landing page [ReadMe](../../../)
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/README.md:20]: Link goes too far up the chain, I think this should be ../../

File: basic_samples/Complex_Dynamic/python/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+request 2.22.0
+
+## Details
+
+See [ReadMe](../) 
+
+### Configuration
+
+Configure desired OMF endpoints to receive the data in [config.ini](.\config.ini]).
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/python/README.md:24]: Extra ending square bracket here, should e (.\config.ini)

File: basic_samples/Complex_Dynamic/python/README.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+password = 
+
+[Configurations]
+omfVersion = 
+compression = 
+WEB_REQUEST_TIMEOUT_SECONDS = 
+VERIFY_SSL = 
+```
+
+note only 1 of OCS, PI, and EDS can be set at a time.
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/python/README.md:46]: Probably capitalize Note

File: basic_samples/Complex_Dynamic/python/README.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+Configuration settings apply to all endpoints.  
+Capitalized fields have default settings that are used if not entered.  
+
+`VERIFY_SSL` may be set to false in order to bypass certificate validation.  This is helpful if PI Web API is configured to use a self-signed certificate. This will generate a warning; this should only be done for testing with a self-signed PI Web API certificate as it is insecure.
+
+#### OSIsoft Cloud Services
+
+An OMF ingress client must be configured.
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/python/README.md:54]: Any docs we can link to on how to do this? I don't think we should list out the steps here but would be nice to point them in the right direction.

File: basic_samples/Complex_Dynamic/python/README.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+## Running the Sample
+
+1. Clone the GitHub repository
+1. Install required modules: `pip install -r requirements.txt`
+1. Open the folder with your favorite IDE
+1. Update `config.ini` with tour credentials
+1. Check and update the program to ensure you are sending to OCS or PI.
+1. Run `program.py` from commandline run `python program.py`
+
+To test the sample after running it:
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/python/README.md:85]: Usually this is under a header ## Running the Automated Tests (or Test, not sure). This is also fine, just slightly inconsistent.

File: basic_samples/Complex_Dynamic/python/auth.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
+
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/python/auth.py:1]: Did you want to run pylint on this sample? I'm trying to start doing that for any of my Python samples, and either resolve the issues or add the comment to the top to ignore them, the Auth PKCE Python sample is an example of that. I can run it and fix those issues if you want.

File: basic_samples/Complex_Dynamic/python/config.ini [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+;For EDS: http://localhost:5590/api/v1/tenants/default/namespaces/default/omf/
+[Access]
+omfURL = 
+
+;This is either your username and password for PI, or clientID and clientSecret for OCS
+[Credentials]
+id = 
+password = 
+
+[Configurations]
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/python/config.ini:20]: I think this should be [Configuration]

File: basic_samples/Complex_Dynamic/python/sendOMF.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        }
+    if config['compression'] == "gzip":
+        msg_headers["compression"] = "gzip"
+
+    authorization = auth.getAuthHeader()
+    if authorization:
+        msg_headers["Authorization"] = authorization
+    return msg_headers
+
+compression
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/python/sendOMF.py:75]: What's this last line about?

File: basic_samples/Complex_Dynamic/python/azure-pipeline.yml [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+          fileLibrary: $(CredFileLibraryOP)
+          filePath: $(projPath)
+
+      - script: |
+          "pip.exe" install -r requirements.txt
+        workingDirectory: '$(Build.SourcesDirectory)$(projPath)'
+        displayName: 'Install requirements'
+
+      - script: |
+          "pytest.exe" ./test_sample.py
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/python/azure-pipeline.yml:90]: If you run this with --junitxml=junit/test-results.xml you can then publish the test results also. Publishing the test results from every pipeline is something I've been trying to add in the Polarity branch, so I can always come back and get that working in that branch.

**Comment by dcbroad3** [comment]
This sample needs to be added to the root README, looks like it has not yet been done in this branch

File: basic_samples/Complex_Dynamic/README.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
+# Complex Dynamic OMF Python Sample
+
+
+| Language | OCS Test Status | EDS Test Status | PI Test Status |
+| -------------- | -------------- | ------------------ |------------------ |
+| <a href="Python3/">Python3</a> | [![Build Status](https://dev.azure.com/osieng/engineering/_apis/build/status/product-readiness/OMF/OMF_DC_Python?branchName=python&jobName=Tests_OCS)](https://dev.azure.com/osieng/engineering/_build/latest?definitionId=1436&branchName=python) | [![Build Status](https://dev.azure.com/osieng/engineering/_apis/build/status/product-readiness/OMF/OMF_DC_Python?branchName=python&jobName=Tests_EDS)](https://dev.azure.com/osieng/engineering/_build/latest?definitionId=1436&branchName=python) |[![Build Status](https://dev.azure.com/osieng/engineering/_apis/build/status/product-readiness/OMF/OMF_DC_Python?branchName=python&jobName=Tests_OnPrem)](https://dev.azure.com/osieng/engineering/_build/latest?definitionId=1436&branchName=python) |
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/README.md:6]: This file still references the folder as "Python3"

File: basic_samples/Complex_Dynamic/python/README.md (hunk 4 above)
Comment (derekendres) [reply basic_samples/Complex_Dynamic/python/README.md]: https://www.youtube.com/watch?v=97QJjUKa6Pk

File: basic_samples/Complex_Dynamic/python/sendOMF.py (hunk 8 above)
Comment (derekendres) [reply basic_samples/Complex_Dynamic/python/sendOMF.py]: no idea... removing

File: basic_samples/Complex_Dynamic/python/README.md (hunk 4 above)
Comment (dcbroad3) [reply basic_samples/Complex_Dynamic/python/README.md]: @derekendres Unless I'm missing something that explains how to create an Authorization Code Client, not how to set up an OMF Ingress Client.

Comment (derekendres) [reply basic_samples/Complex_Dynamic/python/README.md]: the one in the readme was the right link...  https://www.youtube.com/watch?v=52lAnkGC1IM... i apparently copied a different one for in the conversation...

Comment (dcbroad3) [reply basic_samples/Complex_Dynamic/python/README.md]: Yep I see it now

File: basic_samples/Complex_Dynamic/Python/azure-pipeline.yml [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+      - script: |
+          echo Install tools
+          call pip install setuptools wheel
+          call pip install unittest-xml-reporting    
+          call pip install pytest
+          call pip install pytest-cov
+          echo Install requirements
+          call pip install -r requirements.txt
+          echo Run tests
+          call python -m pytest ./test_sample.py --junitxml=junit/test-results.xml
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/Python/azure-pipeline.yml:68]: "call" breaks on Ubuntu systems, this job is passing but logs show it's not actually running any of these commands. Also I see you added the junitxml parameter, but you didn't add a publish test results step.

File: basic_samples/Complex_Dynamic/Python/azure-pipeline.yml [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+          fileLibrary: $(CredFileLibraryOP)
+          filePath: $(projPath)
+
+      - script: |
+          "pip.exe" install -r requirements.txt
+        workingDirectory: "$(Build.SourcesDirectory)$(projPath)"
+        displayName: "Install requirements"
+
+      - script: |
+          "pytest.exe" ./test_sample.py
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/Python/azure-pipeline.yml:90]: This test seems to be failing due to certificate errors

File: basic_samples/Complex_Dynamic/Python/azure-pipeline.yml [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+          fileLibrary: $(CredFileLibraryEDS)
+          filePath: $(projPath)
+
+      - script: |
+          "pip.exe" install -r requirements.txt
+        workingDirectory: "$(Build.SourcesDirectory)$(projPath)"
+        displayName: "Install requirements"
+
+      - script: |
+          "pytest.exe" ./test_sample.py
```
Comment (dcbroad3) [inline basic_samples/Complex_Dynamic/Python/azure-pipeline.yml:111]: This test seems to be failing due to missing omf headers

**Pull request merged**